### PR TITLE
fix/add-tests-for-url-shortner-project

### LINKED ISSUE #37305: URL Shortener Microservice needs tests
I'm creating this as a place to discuss how to create tests for this project, so as not to dominate #37185

> @ojeytonwilliams For URL Shortener, I need to POST URL (google for example) to shorten it and
> it should return `{ "original_url": "www.google.com", "short_url": 1 }`. When I access `$.get(getUserInput(''url'') + ''/api/shorturl/1'')`, I should be redirected to google.com.
> 
> I will need to either 1) store the original_url as a variable OR 2) POST and then GET in one function. Is there a way to do the first one because that would be cleaner and more readable? And where does the first getUserInput get called from actually?
> 
> Here's an excerpt of my test, second test incomplete:
> 
> ```yaml
> tests:
>   - text: 'It should handle a URL as parameter and return shortened URL'
>     testString: 'getUserInput => $.post(getUserInput('url') + '/api/shorturl/new').then(data => assert.exists(data.short_url), xhr => { throw new Error(xhr.responseText)})'
>   - text: 'It should handle a shortened URL and redirect to original link'
>     testString: 'getUserInput => $.get(getUserInput(''url'') + ''/api/shorturl/3'').then()
> ```

### PR DESCRIPTION
**Pull request by abbathaw**
## This PR is closed after a git rebase mistake / disaster. New PR is here #39311 

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution. (I added attribution below).
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #37305

3 of 4 tests have been copied from this [location](https://glitch.com/edit/#!/fcc-back-end-tester?path=public/tests.js:76:236) owned by [christophrus](https://glitch.com/@christophrus) and I was directed to them as per discussions in #37185. 

I went through his tests and found that the example he provided to test the url body using the [fcc-backend-tester](https://glitch.com/~fcc-back-end-tester) was good enough to use and merge here.  

The 4th test is copied from a comment by @ojeytonwilliams in the same issue above.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,6 +3,9 @@
 
 ### VisualStudioCode Patch ###
 
+### Intellij IDEA ###
+/.idea/
+
```
Comment (moT01) [inline .gitignore]: I'm not familiar with this. I assume it won't be an issue, but a little explanation on why it's being added would have been nice.

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md [hunk 2]
```diff
@@ -11,6 +11,15 @@ forumTopicId: 301509
...
 Start this project on Glitch using <a href='https://glitch.com/edit/#!/remix/clone-from-repo?REPO_URL=https://github.com/freeCodeCamp/boilerplate-project-urlshortener/' target='_blank'>this link</a> or clone <a href='https://github.com/freeCodeCamp/boilerplate-project-urlshortener/'>this repository</a> on GitHub! If you use Glitch, remember to save the link to your project somewhere safe!
+
+### Criteria to pass the tests
+
+1. Expose this POST endpoint: `/api/shorturl/new` . 
+      - This endpoint should accept a body with an urlencoded object that has a url key (e.g. `url=https://www.google.com`) . 
+      - It should return a json response with the shortUrl (e.g.`{"original_url":"www.google.com","short_url":1}`).
+1. Return json error message with message of `invalid url` if the url provided in the body of the above POST request isn't a valid URL. e.g. `res.json({"error":"No short url found for given input"})`
+1. Expose this GET endpoint: `/api/shorturl/{:shortUrl}` that will parse the shortUrl given in the params and redirect to the original Url.
+
```
Comment (moT01) [inline curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md]:
```suggestion
```

I like adding instructions here, but they are listed in the project. I think if we want to add the user stories to these backend project challenges, we should make an issue to add them all at once so they are consistent.

**Review by moT01 — CHANGES_REQUESTED** [review]
Thanks for the PR @abbathaw.

The tests looks like they work, I'm not sure if we want to be attaching variables to the window object though. What do you think about that @ojeytonwilliams @RandellDawson?

File: .gitignore (hunk 1 above)
Comment (abbathaw) [reply .gitignore]: this is the settings file generated when using IntelliJ as a code editor. I will remove this.

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md (hunk 2 above)
Comment (abbathaw) [reply curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md]: good point. The user stories are already in the readme file of the project that the learner will clone.

**Comment by RandellDawson** [comment]
@moT01 I do not like the window properties.  @abbathaw Can you not create local variables for the tests?

**Pull request closed without merging**